content/blog/2016-07-23-pull-does-not-scale-or-does-it.md

```diff
@@ -0,0 +1,178 @@
+of Nagios. Nagios has a reputation of not scaling well, in part due to spawning
+subprocesses for active checks that can run arbitrary actions on the Nagios
+host in order to determine the health of a certain host or service. This sort
+of check architecture indeed does not scale well, as the central Nagios host
+quickly gets overwhelmed. As a result, people usually configure checks to only
+be executed every couple of minutes, or they run into more serious problems.
+
+However, Prometheus takes a fundamentally different approach altogether.
+Instead of executing check scripts, it only collects time series data from a
+set of instrumented targets over the network. For each target, the Prometheus
+server simply fetches the current state of all metrics of that target over HTTP
+and has no other execution overhead that would be pull-related. This brings us
+to the next point:
+
+## It doesn't matter who initiates the connection
+
+For scaling purposes, it doesn't matter who initiates the TCP connection over
+which metrics are then transferred. Either way you do it, the effort for
+establishing a connection is small compared to the metrics payload and other
+required work.
+
+But a push-based approach could use UDP and avoid connection establishment
+altogether, you say! True, but the TCP/HTTP overhead in Prometheus is still
+negligible compared to the other work that the Prometheus server has to do to
+ingest data (especially persisting time series data on disk). To put some
+numbers behind this: a single beefy Prometheus server can easily store millions
+of time series, with a record of 800,000 incoming samples per second (as
+measured with real production metrics data at SoundCloud). This allows you to
+monitor over 10,000 machines from a single Prometheus server. The scaling
+bottleneck here has never been related to pulling metrics, but usually to the
+speed at which the Prometheus server can ingest the data into memory and then
+sustainably persist and expire data on disk/SSD.
+
+Also, although networks are pretty reliable these days, using a
+TCP-based pull approach makes doubly sure that metrics data arrives reliably,
+or that the monitoring system at least knows immediately when the metrics
+transfer fails due to a broken network.
+
+## Prometheus is not an event-based system
+
+Some monitoring systems are event-based. That is, they report each individual
+event (an HTTP request, an exception, you name it) to a central monitoring
+system immediately as it happens. This central system then either aggregates
+the events into metrics (StatsD is one example of this) or stores events
+individually for later processing (the ELK stack is an example of that). In
+such a system, pulling would be problematic indeed: the instrumented service
+would have to buffer events between pulls, and the pulls would have to happen
+incredibly frequently in order to simulate the same “liveness” of the
+push-based approach and not overwhelm event buffers.
```

Comment:
This shouldn't make a difference for push vs. pull, same bytes on the wire and you'd need all those buffers anyway to handle the other end being flaky.